### PR TITLE
Add mimetype mapping settings

### LINKED ISSUE #14164: Can't execute APK file downloaded from assets in versions page, on Android
- Gitea version (or commit ref): 1.13
- Operating system: Debian 10
- Database (use `[x]`):
  - [x] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [x] Yes : https://try.gitea.io/poka/test-apk/releases
  - [ ] No

## Description

When download APK android installer from version assets in Gitea **on mobile**, when click the link from APK we can't juste download and execute to install the app.

Instead, browser want to open it with archive unzipper.

If we download the same APK form nextcloud for exemple, Android try to install it directly...

It's not very usefull when we want to share our mobile app from github versions page...

### PR DESCRIPTION
**Pull request by szatyinadam**
This PR should fix #14164 issue.

The `Content-Type` header set to `application/vnd.android.package-archive` instead of `application/zip` when the file extension is .apk . 

Related to #8152 as well.

### CODE REVIEW

File: routers/repo/download.go [hunk 1]
```diff
@@ -59,6 +60,9 @@ func ServeData(ctx *context.Context, name string, size int64, reader io.Reader)
 	} else {
 		ctx.Resp.Header().Set("Content-Disposition", fmt.Sprintf(`attachment; filename="%s"`, name))
 		ctx.Resp.Header().Set("Access-Control-Expose-Headers", "Content-Disposition")
+		if filepath.Ext(name) == ".apk" && base.IsZipFile(buf) {
```
Comment (silverwind) [inline routers/repo/download.go]: Will not work with mixed case letters, e.g. `APK` or `aPk`.

**Comment by silverwind** [comment]
Is there a reason why we single out such types and not just use https://golang.org/pkg/mime/#TypeByExtension?

**Comment by szatyinadam** [comment]
> Is there a reason why we single out such types and not just use https://golang.org/pkg/mime/#TypeByExtension? Are there security implications related to `Content-Type`?

Unfortunately `mime.TypeByExtension(".apk")` returns empty string.

**Comment by silverwind** [comment]
Right, `mime.TypeByExtension` seems worthless. Maybe in the future we can add a more complete mime type database.

File: routers/repo/download.go (hunk 1 above)
Comment (szatyinadam) [reply routers/repo/download.go]: I've fixed this with EqualFold.
Should I squash the fix and the original commit?

**Review by silverwind — APPROVED** [review]

**Comment by szatyinadam** [comment]
Thanks @silverwind for the review! Can someone else check it?

**Comment by zeripath** [comment]
I think we just need to think carefully about whether there are security implications of this change.

**Comment by silverwind** [comment]
Indeed, it is related to https://github.com/go-gitea/gitea/issues/8152#issuecomment-814351404 and recent PR https://github.com/go-gitea/gitea/pull/15299.

Github [serves raw APK](https://github.com/AppsFlyerSDK/appsflyer-android-app/blob/master/app/release/app-release.apk) as generic mime type `application/octet-stream`.

I now think we should in the default configuration serve generic mime types `application/octet-stream` and `text/plain` to be on the safe side. There could be added a configuration option to serve "more correct" mime types, maybe user-definable based on file extension.

**Comment by silverwind** [comment]
See https://github.com/go-gitea/gitea/issues/8152#issuecomment-819724255 for a better suggestion.

**Comment by szatyinadam** [comment]
@silverwind Can You take an other look for the new commit? If the direction is good then I will remove the unused functions created in the previous commit and remove the Draft flag.

**Comment by silverwind** [comment]
Looks good at first glance. Yes, please make it case-insenstive. Should add an entry in https://github.com/go-gitea/gitea/blob/master/docs/content/doc/advanced/config-cheat-sheet.en-us.md as well.

I'm not totally satisfied with the name of the config option, maybe you find a better section to put it in or someone else has a better suggestion.

**Comment by silverwind** [comment]
Maybe we can put it as `security.mimetype.mapping` as it is kind of a security-related topic.

**Comment by szatyinadam** [comment]
@silverwind Can you take a look on this PR or refer anyone who can do a review?

File: custom/conf/app.example.ini [hunk 2]
```diff
@@ -1302,3 +1302,8 @@ STORAGE_TYPE = local
 ;MINIO_LOCATION = us-east-1
 ; Minio enabled ssl only available when STORAGE_TYPE is `minio`
 ;MINIO_USE_SSL = false
+
+; Custom mime type mapping for downloadable files
+;[download.mimetype.mapping]
```
Comment (zeripath) [inline custom/conf/app.example.ini]:
```suggestion
;[repository.mimetype_mapping]
```

File: docs/content/doc/advanced/config-cheat-sheet.en-us.md [hunk 3]
```diff
@@ -958,6 +958,15 @@ MINIO_USE_SSL = false
 
 And used by `[attachment]`, `[lfs]` and etc. as `STORAGE_TYPE`.
 
+## MIME type mapping (`download.mimetype.mapping`)
```
Comment (zeripath) [inline docs/content/doc/advanced/config-cheat-sheet.en-us.md]: Move it up to near the repository sections and rename it:

```suggestion
## MIME type mapping (`repository.mimetype_mapping`)
```

File: modules/setting/mime_type_map.go [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+		Enabled bool
+		Map     map[string]string
+	}{
+		Enabled: false,
+		Map:     map[string]string{},
+	}
+)
+
+func newMimeTypeMap() {
+	sec := Cfg.Section("download.mimetype.mapping")
```
Comment (zeripath) [inline modules/setting/mime_type_map.go]:
```suggestion
	sec := Cfg.Section("repository.mimetype_mapping")
```

**Review by zeripath — COMMENTED** [review]
I think changing the name to repository.mimetype_mapping would be better

File: docs/content/doc/advanced/config-cheat-sheet.en-us.md [hunk 5]
```diff
@@ -958,6 +958,15 @@ MINIO_USE_SSL = false
...
+
+Configuration for set the expected MIME type based on file extensions of downloadable files. Configuration presents in key-value pairs and file extensions starts with leading `.`.
+
+The following configuration set `Content-Type: application/javascript` header when downloading files with `.js` file extension.
```
Comment (techknowlogick) [inline docs/content/doc/advanced/config-cheat-sheet.en-us.md]: I'm hesitant to have js be the example(same for .ini as well), could you have it be just apk? As JS can be executed by the browesr, so I don't want people copy pasting things without realizing what they are configuring and opening themselves up to security issues.

File: custom/conf/app.example.ini [hunk 6]
```diff
@@ -1302,3 +1302,8 @@ STORAGE_TYPE = local
...
+;.apk=application/vnd.android.package-archive
+;.js=application/javascript
```
Comment (techknowlogick) [inline custom/conf/app.example.ini]: same as my other note

**Review by zeripath — APPROVED** [review]

**Review by techknowlogick — APPROVED** [review]
Thanks for this PR :)

**Pull request merged**